### PR TITLE
[asmjit] Update to latest commit

### PR DESCRIPTION
**Pull request by Chlumsky**
On the anniversary of the previous version update, I submit this one.

- #### What does your PR fix?
  It updates the AsmJit package, notably fixing asmjit/asmjit#380

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  No changes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Sure.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Sure.

### CODE REVIEW

File: ports/asmjit/vcpkg.json [hunk 1]
```diff
@@ -1,7 +1,6 @@
 {
   "name": "asmjit",
-  "version-date": "2021-10-26",
-  "port-version": 1,
+  "version-date": "2022-10-26",
   "description": "Complete x86/x64 JIT and Remote Assembler for C++",
   "homepage": "https://github.com/asmjit/asmjit",
```
Comment (MonicaLiu0311) [inline ports/asmjit/vcpkg.json:5]:
```suggestion
  "homepage": "https://github.com/asmjit/asmjit",
  "license": "Zlib",
```

Comment (MonicaLiu0311) [reply ports/asmjit/vcpkg.json:5]: A license is required.

Comment (Chlumsky) [reply ports/asmjit/vcpkg.json:5]: But it wasn't required before?

**Comment by MonicaLiu0311** [comment]
@Chlumsky 
CI tests fail on many triplets, Please get [failure logs here](https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/80027/artifacts?artifactName=failure%20logs%20for%20x86-windows&api-version=7.0&%24format=zip). Please ping me if you need any help.

**Comment by Chlumsky** [comment]
Asmjit itself works fine, but two other packages, `fbgemm` and `polyhook2` are not compatible with the new version of Asmjit, but I believe they would be if they were updated also.

**Comment by stevemk14ebr** [comment]
It's a messy situation that we're stuck in. Vcpkg doesn't have version control so we can't get out of the issue of combinations of asmjit, asmtk, FBGEMMs, and polyhook2 depending on differing versions.

The blocker right now is FBGEMMs. Their cmakelist relies on a patch file that doesn't work with their master. Their master is the latest version that supports the new asmjit. Polyhook2 master already supports the new asmjit. 

I've been to their issue tracker a few times about this but nothing fixed yet. I've considered submitting a PR that removes FBGEMM - at least temporarily and then updates asmjit, polyhook2, and adds asmtk all at once. This would fix the current situation

The correct solution is to add versioning to vcpkg but yea.

**Comment by MonicaLiu0311** [comment]
As you already know, vcpkg does not support version control. If you've patched/updated downstream usage, their new version supports the latest asmjit, you can update together in this PR.

**Comment by MonicaLiu0311** [comment]
@BillyONeal 
Do you have any good advice?(Last related PR: https://github.com/microsoft/vcpkg/pull/24362)

**Comment by stevemk14ebr** [comment]
I can push an update for asmjit, asmtk, and polyhook2 but it would require removing fbgemm as their cmakelist/build is broken

**Comment by BillyONeal** [comment]
> It's a messy situation that we're stuck in. Vcpkg doesn't have version control so we can't get out of the issue of combinations of asmjit, asmtk, FBGEMMs, and polyhook2 depending on differing versions.

Strictly speaking, this constraint does not come from vcpkg, it comes from C++. A program can only have 1 definition of a symbol, which means you can't load multiple versions of a component at the same time. vcpkg is forwarding this constraint to you in how we expose packages.

We do have precedent for updating a shared dependency of a lot of stuff even when not everyone can take the update; for example, we turned off `usd` in order to update `tbb`.

>I've been to their issue tracker a few times about this but nothing fixed yet. I've considered submitting a PR that removes FBGEMM - at least temporarily and then updates asmjit, polyhook2, and adds asmtk all at once. This would fix the current situation

Is this the sort of thing like with usd and tbb where usd is used ~never in comparison with how tbb is used (in this case fbgemm extremely rarely used in comparison to these other ports)?

**Comment by stevemk14ebr** [comment]
> Strictly speaking, this constraint does not come from vcpkg, it comes from C++. A program can only have 1 definition of a symbol, which means you can't load multiple versions of a component at the same time. vcpkg is forwarding this constraint to you in how we expose packages.

I don't think the issue at hand is quite the same as the C++ issue. Each library in question only links one version of asmjit. The issue is each library requires a _different_ version of asmjit than each other library. So vcpkg _could_ have a versioning system where each library said 'hey I want version XXX' and that's what gets linked. There's no try to link different versions multiple times or anything like that here.

> Is this the sort of thing like with usd and tbb where usd is used ~never in comparison with how tbb is used (in this case fbgemm extremely rarely used in comparison to these other ports)?

I am not actually familiar with fbgemm at all. About all I can say is their builds have been broken for quite a while and I've not seen (m?)any complaints from users about this. Meanwhile, asmjit is heavily used. And I can say I've had multiple users ask me when polyhook's vcpkg will be fixed.

My personal preference (biased) is we rip out fbgemm from vcpkg, fix the other packages. And if they care enough to fix their build, they can add it back. At least it unblocks this mess.

**Comment by BillyONeal** [comment]
>Each library in question only links one version of asmjit. The issue is each library requires a different version of asmjit than each other library.

C++ does not let you create this situation; that breaks the ODR.

>So vcpkg could have a versioning system where each library said 'hey I want version XXX' and that's what gets linked.

No, C++ does not let you create such a situation. At runtime there can only be one provider of a given symbol in the process. (Yes, on Windows, if you static link everything and don't expose the dependency from a binary, sometimes you might possibly be able to get away with it. On all other platforms you can't even in that case because the operating system loader says no)

>What you're saying would be true if say fbgemm used polyhook as well as asmjit, and the versions of asmjit used by polyhook and fbgemm differed. There would be a collision on the transitive asmjit dependency in that case.

If any end user wants fbgemm and polyhook, and they required different versions of asmjit, the end user's program no longer links. There is no combination to put on the link line that will work since the linker is only going to tolerate one asmjit.

>My personal preference (biased) is we rip out fbgemm from vcpkg, fix the other packages. And if they care enough to fix their build, they can add it back. At least it unblocks this mess.

OK let me ask what other maintainers think.

**Comment by stevemk14ebr** [comment]
Again, I think we're talking about two different things. I'm not talking about linking or ODR or anything C++ related at all. I'm talking about libraries and the version of the dependencies they use.

> C++ does not let you create this situation; that breaks the ODR.
As it is right now, fbgemm and polyhook are free to use differing versions of asmjit. Both will be built, and linked, and work just fine. The issue at hand is that vcpkg requires there to be _one_ version of asmjit. So polyhook and fbgemm cannot coexist in vcpkg at the same time as vcpkg is not advanced enough to allow each library to specify what version of asmjit each one need individually.

**Comment by BillyONeal** [comment]
> I'm talking about libraries and the version of the dependencies they use.

Libraries are not islands. All their symbols are visible at link time.

>As it is right now, fbgemm and polyhook are free to use differing versions of asmjit. Both will be built, and linked, and work just fine.

They will almost certainly actually get some mishmash of symbols from each version because that's how linking works.

**Comment by stevemk14ebr** [comment]
We differ because of this assumption

> If any end user wants fbgemm and polyhook, and they required different versions of asmjit, the end user's program no longer links. There is no combination to put on the link line that will work since the linker is only going to tolerate one asmjit.

It's pretty strong to assume that a downstream user wants N # of vcpkg libraries to all work together. I would never use fbgemm, why should polyhook have to be version compatible.

**Comment by BillyONeal** [comment]
@ras0219-msft @JavierMatosD @markle11m @vicroms @AugP @valeriaconde @dan-shaw, and I
 discussed this and are OK with marking fbgemm as a failure if and only if it is impossible to update it in a consistent way. @ras0219-msft notes that fbgemm made a release 6 days ago that may work.

**Comment by stevemk14ebr** [comment]
The issue is their cmakelists relies on an outdated patchfile, unfortunately.

**Comment by MonicaLiu0311** [comment]
Close this PR as it has been moved to https://github.com/microsoft/vcpkg/pull/27683.

**Pull request closed without merging**